Add tests for REST API authenticated by token in parameter

olivergondza opened the pull request:
I suspect that the [advertised](https://wiki.jenkins-ci.org/display/JENKINS/Remote+access+API)  way of REST authentication is broken: `curl -X POST <JENKINS_URL>/job/<JOB_NAME>/build --data token=<TOKEN>`. I have created a couple of reproducing tests and it seems to confirm the hypothesis that the token parameter just does not authenticate the user.

@jenkinsci/code-reviewers, can you prove or disprove this? I am fine if that does not work but fixing the docs is the minimum.

olivergondza committed:
Add tests for REST API authenticated by token

daniel-beck commented:
> where TOKEN is set up in the job configuration

This has nothing to do with the API token for a user, but is a token for the "Trigger builds remotely" pseudo-trigger. It is only useful if anon (or your actual Basic auth user/pass) have Job/Read, but not Job/Build. It's probably from the time when Jenkins permissions were *Logged in users can do anything* (and anon could see anything).

Also, that wiki page is so wrong it's not even funny. Nothing on there about triggering parameterized builds is correct. They just pretend they're a web browser submitting the (GET) `/buildWithParameters` form rather than use the actual `POST /buildWithParameters` method documented on `/job/foo/api`.

olivergondza commented:
Oh, this confused me the same way it confused several of our users. I will update the tests to cover that use case as well and hopefully move the fixed remote API page to jenkins.io.

olivergondza commented:
Alight, I have removed the most confusing thing from the wiki and removed the test that does no longer makes. As the token trigger is deprecated, anyway.

olivergondza commented:
Does not seem related to me.

```
[ERROR] npm ERR! Windows_NT 6.3.9600
[ERROR] npm ERR! argv "C:\\Packages\\Plugins\\Microsoft.Compute.CustomScriptExtension\\1.8\\Downloads\\0\\workspace\\Core_jenkins_PR-2768-FCKU2NXOI6ZERWDW67GWYJUHDLS6OTDM6RL6LJBMGR2BQUZJEW2Q\\war\\node\\node.exe" "C:\\Packages\\Plugins\\Microsoft.Compute.CustomScriptExtension\\1.8\\Downloads\\0\\workspace\\Core_jenkins_PR-2768-FCKU2NXOI6ZERWDW67GWYJUHDLS6OTDM6RL6LJBMGR2BQUZJEW2Q\\war\\node\\node_modules\\npm\\bin\\npm-cli.js" "install"
[ERROR] npm ERR! node v4.0.0
[ERROR] npm ERR! npm  v3.8.9
[ERROR] 
[ERROR] npm ERR! Callback called more than once.
[ERROR] npm ERR! 
[ERROR] npm ERR! If you need help, you may report this error at:
[ERROR] npm ERR!     <https://github.com/npm/npm/issues>
[ERROR] 
[ERROR] npm ERR! Please include the following file with any support request:
[ERROR] npm ERR!     C:\Packages\Plugins\Microsoft.Compute.CustomScriptExtension\1.8\Downloads\0\workspace\Core_jenkins_PR-2768-FCKU2NXOI6ZERWDW67GWYJUHDLS6OTDM6RL6LJBMGR2BQUZJEW2Q\war\npm-debug.log
```

olivergondza committed:
Merge branch 'master' into token-tests

olivergondza commented:
@daniel-beck, @oleg-nenashev, can we merge this?

daniel-beck commented on test/src/test/java/jenkins/security/ApiTokenPropertyTest.java:
```diff
@@ -116,17 +123,38 @@ public void adminsShouldBeUnableToChangeTokensByDefault() throws Exception {
+
+        WebClient wc = createClientForUser("foo");
+        WebRequest wr = new WebRequest(new URL(j.getURL(), "job/bar/build"), HttpMethod.POST);
+        wc.addCrumb(wr);
```
This should be unnecessary since 2.96 or so.

olivergondza replied:
You are right, as always. Amended.

olivergondza committed:
Merge branch 'master' into token-tests

olivergondza committed:
Avoid adding the crumb that is not needed anymore

oleg-nenashev reviewed (approved):
LGTM. CC @Wadeck who was working on this area recently

jglick commented:
`BuildAuthorizationToken` has long been deprecated but without any compelling alternative, since Jenkins lacks any system for tokens with limited scope. This is filed in JIRA by me with discussion somewhere.
